### ISSUE TITLE
perf: PHOTONモデルgenerate()にKVキャッシュを実装して推論速度を改善

### ISSUE THREAD

**Issue by Kewton** [Kewton/photon-mlx#54]
## 背景・課題

現在の `photon_mlx/model.py` の `generate()` は、1トークン生成するたびに全シーケンスのフォワードパスを実行している。

```python
# 現在の実装 (model.py:309)
for _step in range(max_new_tokens):
    logits, _loss = self.__call__(ids, labels=None)  # 毎ステップ全体を bottom-up → top-down で再計算
```

実際には毎ステップ `padded_len = pad_to_multiple(T + max_new_tokens, total_chunk)` 長のシーケンス全体を bottom-up → top-down で再計算しており、max_new_tokens 回の再計算が冗長である。具体的には 1 ステップあたりの実コストは `O(T + (T/CS[0])² + (T/∏CS)²)`（階層エンコーダの二乗項が支配）であり、768 トークン生成する場合はそれが768回実行される。

なお「最終ステップでは最大 T+768 トークンを処理する」という表現は正確ではなく、`pad_to_multiple` によって **毎ステップ同じ padded_len** が処理される。

## 実装可能性の分析

PHOTONの階層構造を考慮した KV キャッシュの設計：

### アーキテクチャの特性

- **ローカルデコーダ**（`_decode_level`）: super-chunk 単位（L = `CS[0] * (PL[0] + 1)`、デフォルト config で 12）でアテンション → super-chunk 内のサイズが小さいため KVキャッシュ不要。ただし新トークン追加により末尾 super-chunk の再計算は毎ステップ発生することに注意。
- **Level-1 エンコーダ**（デフォルト config で T/4 トークン、一般化: T/CS[0]）: 新トークンは最後のチャンクのみ影響 → 差分更新可能
- **Level-2 エンコーダ**（デフォルト config で T/16 トークン、一般化: T/∏CS）: 同上 → 差分更新可能
- **トップレベルデコーダ**（デフォルト config で T/16 トークン）: 安定した prefix 部分には KVキャッシュが適用可能。ただし末尾の進行中 super-chunk に対応する最後の top-level 位置は新トークンごとに hidden state が変化するため、append-only の標準 KV cache ではなく tail slot の再計算・差し替えを前提に設計する必要がある。

#### Phase 1: トップレベル KV キャッシュ（優先度: 高）

`blocks.py` の `Attention` クラスにKV引数を追加し、`generate()` のトップレベルアテンション（T/16シーケンス）でキャッシュを利用する。

```python
# blocks.py: Attention.__call__ にKVキャッシュ引数・RoPE位置オフセット引数を追加
def __call__(self, x, cos, sin, mask=None, kv_cache=None, position_offset: int = 0):
    # kv_cache: (K_prev, V_prev) | None
    # RoPE: cos[offset:offset+T] を参照する（_apply_rope も改修が必要）
    # 新トークンのK,Vを計算し kv_cache に concat / replace
```

**RoPE 位置オフセット対応**: `blocks.py:22-32` の `_apply_rope` は現状 `cos[:T]` と先頭 T 要素のみを用いる。KVキャッシュ導入後に新トークン 1 つだけを処理する場合、その位置は `T_prev` であるため `cos[T_prev:T_prev+1]` を適用する必要がある。`Attention.__call__` に `position_offset: int = 0` 引数を追加し、`_apply_rope` も `cos[offset:offset+T]` を参照するよう改修する。これを怠ると受入条件「logits が一致」を満たせない。なお top-level decoder では `position_offset` は raw token 長ではなく **top-level sequence position** 基準で扱い、末尾 tail slot を再計算する間は offset を進めず、次の top-level 位置が実際に追加されるタイミングでのみ進める。

**後方互換性**: `kv_cache=None`, `position_offset=0` をデフォルトとし、既存の呼び出し箇所（model.py:202, 233, 242, inference.py:80, 87）がデフォルト引数で問題なく動くことを保証する。また `generate()` に `use_kv_cache: bool = True` フラグを追加し、`use_kv_cache=False` で現状と同一コードパスを残してトラブル時に切り替えられるようにする。`use_kv_cache` は `generate()` の公開 API 用フラグに留め、`Attention` / `TransformerBlock` には追加せず、内部詳細は `kv_cache is None` で分岐する。

**呼び出し箇所マッピング（Phase 1）**: `Attention.__call__` および `TransformerBlock.__call__` のシグネチャ変更により、以下の全 5 呼び出し箇所が影響を受ける。**Phase 1 で `kv_cache` を実際に渡す箇所はトップレベルデコーダのみ**（model.py:242 / inference.py:87）。エンコーダ（model.py:233 / inference.py:80）・ローカルデコーダ（model.py:202）は `kv_cache=None` のままとし、blocks.py:128 の `TransformerBlock.__call__` 内部 `self.attn(...)` についても同様にデフォルト引数でキャッシュ不要パスを維持する。誤って全呼び出し箇所に一律キャッシュを渡す実装を防ぐため、本マッピングを実装の起点とすること。

**KVキャッシュのスコープ（Phase 1）**: Phase 1 では `generate()` 内の一時キャッシュ（関数スコープのローカル変数）とする。Phase 2 で session 跨ぎの事前計算 prefix キャッシュへ拡張する設計とする。`HierarchicalState`（`session.py:17-24`）への格納は Phase 2 で検討する。

**KVキャッシュのデータ構造（Phase 1）**: `Attention.__call__` が受け取る 1 layer 分の `kv_cache` は `(k, v)` tuple のままでよいが、`generate()` 内部の保持構造はトップレベル decoder layer 数に対応した ordered list とし、`top_level_kv_cache: list[tuple[mx.array, mx.array] | None]` を `self.decoders[L - 1]` の layer 順に保持する。`for layer_idx, block in enumerate(self.decoders[L - 1])` と 1:1 対応させ、保存する K/V は `mx.repeat` 前の `n_kv_heads` 形状に統一する。Phase 1 では dataclass 化まで行わず、layer index が自明な list で十分とする。なおこの layer-indexed list 自体の追加メモリは軽微であり、Phase 1 の 50 MB 上限を支配するのは list オーバーヘッドではなく各 layer の K/V tensor 本体である。

**Phase 1 の tail 再計算範囲**: Phase 1 は「トップレベルデコーダだけを更新する」では足りない。トップレベルデコーダの入力 `enc_outputs[L]` 自体が新トークン追加で末尾側だけ変化するため、increment ごとに少なくとも `token_proj` → level-1 chunker の末尾入力 → level-1 encoder → level-2 chunker の末尾入力 → level-2 encoder を経てトップレベル入力の末尾 1 位置を更新する必要がある。**重要**: Stage 5 の呼び出し箇所マッピングでは Phase 1 で encoder 側に `kv_cache` を渡さないため、局所化できるのは chunker 入力の組み立てまでであり、`encoders[0]` / `encoders[1]` 自体は causal self-attention の性質上、各レベルの compressed sequence 全体（デフォルトで level-1 は T/4、level-2 は T/16）を毎ステップ replay する前提とする。KVキャッシュで再利用できるのは **トップレベル decoder の安定した prefix 位置** のみであり、進行中 super-chunk に対応する最後の top-level 位置は `actual_len % prod(chunk_sizes) == 0` で次スロットが確定するまで毎ステップ再計算・cache 差し替え対象とする。

**プリフィルと増分デコードの分岐**: プリフィル時は `prompt_len` まで通常の bottom-up 処理を実行し、以降の increment では (a) 影響する bottom-up tail 経路の再計算、(b) トップレベル decoder の stable prefix cache 再利用 + tail slot 差し替え、(c) local_decoder の末尾 super-chunk 再計算、を行う。ここで (a) の encoder replay コストが Phase 1 に残る主要ボトルネック候補であり、T=2048 では 2 倍目標を厳しくする可能性があるため、ベンチマークで必ず内訳を分離して計測する。これにより `pad_to_multiple` で確保した padded バッファが増分フォワードでは意味を持たなくなるため、増分パスでの padding 処理を整理する。

**increment path の mask / cache replace ルール**: top-level increment では query 長 1、key 長は `stable prefix + current tail` となる。このため増分パスは `causal_mask(T)` をそのまま流用する full-sequence 経路とは分け、`Attention.__call__` は `mask=None` もしくは rectangular mask を許容する前提にする。また `kv_cache` は append-only ではなく、進行中 super-chunk に対応する最終 slot を replace できる API として扱う。RoPE offset・mask・cache 更新規則は `blocks.py` 側の責務として一体で明記する。

#### Phase 2: 階層全体の差分更新（優先度: 中）

`HierarchicalState`（`session.py`）を拡張し、エンコーダ出力のチャンク単位差分更新を実装する。新トークンが追加されるたびに末尾チャンクのみ再計算する。Phase 2 では KVキャッシュを session 跨ぎの事前計算 prefix キャッシュとして `HierarchicalState` に格納する設計を検討する。

### 期待効果

| コンテキスト長 | 現在（全再計算） | KVキャッシュ後 | 単位 |
|--------------|----------------|--------------|-----|
| T=2048, gen=768 | O(2048 + (512)² + (128)²) × 768 ≈ 高コスト | プリフィル1回 + 増分 × 768（top-level cache + encoder replay 残存） | アテンション行列要素数 (proxy) |
| T=32768, gen=768 | O(32768 + (8192)² + (2048)²) × 768 ≈ 非常に高コスト | プリフィル1回 + 増分 × 768（top-level cache + encoder replay 残存） | アテンション行列要素数 (proxy) |

トップレベル（デフォルト config で T/16）のみのキャッシュでも推論速度が **5〜15倍** 向上する見込み、という当初想定は encoder replay 残存前の楽観値である。Stage 5/7 の tail recompute 設計を踏まえると、Phase 1 の実効 speedup は `top-level cache + local tail 再計算 - encoder replay 残存コスト` で決まり、特に T=2048 では 2 倍未達の可能性がある。したがって 5〜15倍は達成必須ではなく参考値に留め、受入条件の 2 倍判定はベンチ内訳と併せて評価する。

## 関連ファイル

- `photon_mlx/blocks.py`: `Attention.__call__` — KVキャッシュ引数・`position_offset` 追加、`_apply_rope` の offset 対応
- `photon_mlx/model.py`: `PhotonModel.generate()` — キャッシュを使ったデコードループ実装（`use_kv_cache` フラグ追加）、`TransformerBlock.__call__` の kv_cache パススルー対応。加えて `__call__` / `hierarchical_prefill()` / generate prefill が共有する `_encode_bottom_up()` / `_decode_from_enc_outputs()` 相当の helper 抽出候補をここに集約する。
- `photon_mlx/session.py`: `HierarchicalState` — Phase 2 で KVキャッシュ保持用フィールド追加（Phase 1 は generate() 内一時変数で対応）
- `photon_mlx/inference.py`: `hierarchical_prefill()` — `Attention` シグネチャ変更への追随が必要。なお `hierarchical_prefill()` は `PhotonModel.__call__` と同一の bottom-up → top-down 処理を再実装しており（コード重複の技術的負債）、tail recompute 設計を入れると generate prefill を含めて3経路に増える。shared helper 化を前提に影響範囲を閉じる。
- `photon_mlx/tests/test_generate.py`: KVキャッシュ導入で最も影響を受けるテストファイル。既存テストクラスに加え、`TestKVCacheEquivalence` クラスを新規追加し `use_kv_cache=True/False` の等価性を恒久テストとして組み込む（後述）。
- `bench/kv_cache_speedup.py`: Phase 1 完了後の tokens/sec 比較ベンチマーク（正式化）。`reset_peak_memory()` → warmup → N 回 generate → `measure_memory()` のシーケンスで速度・メモリ双方を計測する自己完結スクリプト。既存の `benchmark_forward`（optimize.py:129）は forward-only 計測のため generate() 全体の tokens/sec 計測には転用不可。
- `scripts/generate_photon.py`: line 131 の `print('(KV cache not implemented — latency is expected)')` を Phase 1 完了時に削除または `'KV cache enabled (Phase 1, top-level only)'` 等に書き換えること。また `--no-kv-cache` CLI フラグ追加（オプション）を Phase 1 完了時タスクとして検討する。
- **影響なし（baseline_reporag）**: `baseline_reporag/photon_pipeline.py` は `PhotonModel.generate()` を使用せず、`hierarchical_prefill()`（line 342, 451）のみを経由する（line 484 の `bl.generator.generate(...)` は `baseline_reporag/generation/generator.py` の mlx_lm で別物）。`Attention` シグネチャ変更の後方互換（`kv_cache=None` デフォルト）を保てば `PhotonRAGPipeline` / Safe RecGen の挙動は完全に不変。

### Phase 1 成功基準

- [ ] `generate(use_kv_cache=True)` と `generate(use_kv_cache=False)` の greedy decode における next_token 列が完全一致すること
- [ ] `generate(use_kv_cache=True)` と `generate(use_kv_cache=False)` の logits が許容誤差 `atol=1e-5, rtol=1e-4` 以内で一致すること（MLX の浮動小数点演算順序の差異を考慮し bitwise 一致は必須としない）
- [ ] T=2048, max_new_tokens=64 で推論速度が現在比 **2倍以上** 向上すること（tokens/sec で計測、`bench/kv_cache_speedup.py` を使用）
- [ ] `bench/kv_cache_speedup.py` は少なくとも `prefill`, `encoder_replay`, `top_level_increment`, `local_tail_decode` の内訳を出力し、Phase 1 の残存ボトルネックを説明できること
- [ ] T=2048 の条件で KVキャッシュ追加によるメモリ増加量が **50 MB 以下** であること（Phase 1 top-level decoder のみキャッシュ時の実測値。Phase 2 階層全体キャッシュ含む上限は 500 MB）。`photon_mlx/optimize.py` の `reset_peak_memory()` → generate → `measure_memory()` のシーケンスで計測する。
- [ ] **Phase 1 単独で 2 倍未達の場合**: `bench/kv_cache_speedup.py` で計測した実測結果と Phase 2 着手判断レポートを本 Issue に残し、Phase 2 別 Issue 継続の判断基準とする（= Phase 1 単独では Issue #54 クローズ不可）。
- [ ] Phase 1 完了時に `bench/` 配下でプロファイルを取得し、トップレベルデコーダが実際のボトルネックであることを確認すること（そうでなければ Phase 2 の優先度を引き上げる）
- [ ] `python -m pytest torch_ref/tests/ photon_mlx/tests/ -v` が全 72 テストパスすること。特に以下のファイルが回帰しないことを確認する:
  - `photon_mlx/tests/test_generate.py`: `TestGenerateShape`, `TestGenerateDeterminism`, `TestGeneratePrompt`, `TestGenerateBoundary`, `TestGenerateQuality` および新規 `TestKVCacheEquivalence`（後述）
  - `photon_mlx/tests/test_forward.py`: `TestShapes`, `TestChunkBoundary`, `TestSmoke`（`TransformerBlock` を通過する全テスト）
  - `photon_mlx/tests/test_session.py`: `TestInference`（`hierarchical_prefill` 経由）、`TestPruneEvidence`
  - `photon_mlx/tests/test_training.py`: 特に `TestLoadModel::test_load_model_forward_consistency`（`mx.allclose(..., atol=1e-5)` による数値等価性）— `kv_cache=None` デフォルトの `__call__` 経路で before/after 一致を維持すること
  - `photon_mlx/tests/test_optimize.py`: `TestWarmup`, `TestBenchmark`
- [ ] `photon_mlx/tests/test_generate.py` に `TestKVCacheEquivalence` クラスを追加し、少なくとも以下のメソッドを恒久テストとして組み込むこと:
  - `test_step_logits_match_without_cache`: `use_kv_cache=True/False` の step logits が `atol=1e-5, rtol=1e-4` で一致すること
  - `test_cross_level1_boundary`: `prompt_len=15, max_new_tokens=2` と `prompt_len=16, max_new_tokens=1` を用い、4-token chunk 境界で tail overwrite が正しく動くこと
  - `test_cross_top_level_boundary`: `prompt_len=31, max_new_tokens=2` と `prompt_len=32, max_new_tokens=1` を用い、16-token super-chunk 境界で top-level tail slot の replace / append 切替が正しく動くこと
  - `test_multiple_prompt_lengths`: `prompt_len=16 / 17 / 32 / 64` で greedy token 列一致を確認すること
- [ ] 追加する境界跨ぎテストは既存 `model` fixture を再利用し、小サイズ prompt の parametrized case に留めて CI 時間増を最小化すること（`test_forward.py` / `test_session.py` は主として `kv_cache=None` デフォルト経路の回帰確認先とする）
- [ ] `ruff check .` / `ruff format --check .` がパスすること（警告0件）

### Phase 2 成功基準（参考）

- [ ] T=2048, max_new_tokens=64 で推論速度が現在比 **5倍以上** 向上すること（Phase 2 完了後に計測）
- [ ] session 跨ぎの prefix キャッシュが `HierarchicalState` に正しく格納・復元されること

## 将来検討

- **Phase 1 / Phase 2 の Issue 分割**: Phase 2（階層全体の差分更新）は規模が大きく優先度が中のため、別 Issue として切り出すことを検討する。リリース単位でのPR分割にも寄与する。
- **バッチ次元（B>1）・変動 prompt 長への対応**: 現状の `generate()` は `input_ids.shape = (B, T_prompt)` を受け取り B>1 に対応。KVキャッシュ導入時は `pad_input_ids`（`photon_mlx/optimize.py`）で prompt 長を揃えている前提で設計し、B>1 で prompt 長が異なる場合の padding の扱いを明記する。
- **ベンチマークハーネスの整備**: 受入条件の「推論速度2倍向上」を定量的に確認するため `bench/kv_cache_speedup.py`（before/after の tokens/sec 比較スクリプト）を追加する（関連ファイル節で正式化）。
- **PHOTON モジュール docstring 更新**: `photon_mlx/model.py:1-14` のモジュール docstring に「推論時 KV キャッシュパス（Phase 1: top-level decoder のみ増分更新）」の概要を追記する（Phase 1 完了後に対応）。
- **TurnProfiler への拡張**: `baseline_reporag/profiler.py` の `TurnProfiler` に `photon_prefill` / `photon_increment` フェーズを追加して PHOTON 内部のプロファイリングを強化する（Phase 1 後の別 Issue 候補）。
